### ISSUE TITLE
Widget killed my dashboard

### ISSUE THREAD

**Issue by bossanova808** [putyourlightson/craft-campaign#396]
I was just checking out the Campaign dashboard widget - I got an error 'couldn't save widget' and now on all Dashboard page loads, I am getting:
```
Integrity constraint violation – [yii\db\IntegrityException](https://www.yiiframework.com/doc-2.0/yii-db-integrityexception.html)
SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'dateCreated' in where clause is ambiguous
The SQL being executed was: SELECT COUNT(*)
FROM (SELECT `campaign_sendouts`.`lastSent` AS `lastSent`, `elements`.`id` AS `elementsId`, `elements_sites`.`id` AS `elementsSitesId`, `content`.`id` AS `contentId`
FROM `craft_elements` `elements`
INNER JOIN `craft_campaign_campaigns` `campaign_campaigns` ON `campaign_campaigns`.`id` = `elements`.`id`
LEFT JOIN (SELECT `campaignId`, MAX(`lastSent`) AS `lastSent`
FROM `craft_campaign_sendouts`
GROUP BY `campaignId`) `campaign_sendouts` ON `campaign_sendouts`.`campaignId` = `campaign_campaigns`.`id`
INNER JOIN `craft_campaign_campaigntypes` `campaign_campaigntypes` ON `campaign_campaigntypes`.`id` = `campaign_campaigns`.`campaignTypeId`
INNER JOIN `craft_sites` `sites` ON `sites`.`id` = `campaign_campaigntypes`.`siteId`
INNER JOIN `craft_elements_sites` `elements_sites` ON `elements_sites`.`elementId` = `elements`.`id`
INNER JOIN `craft_content` `content` ON `content`.`elementId` = `elements`.`id`
WHERE (`campaign_campaigns`.`campaignTypeId`=2) AND (`sites`.`dateDeleted` IS NULL) AND ((`dateCreated` >= '2023-03-13 13:00:00') AND (`dateCreated` < '2023-06-12 04:16:13')) AND (`elements`.`archived`=FALSE) AND ((`elements`.`enabled`=TRUE) AND (`elements_sites`.`enabled`=TRUE)) AND (`elements`.`dateDeleted` IS NULL) AND (`elements`.`draftId` IS NULL) AND (`elements`.`revisionId` IS NULL)) `subquery`
INNER JOIN `craft_elements` `elements` ON `elements`.`id` = `subquery`.`elementsId`
INNER JOIN `craft_elements_sites` `elements_sites` ON `elements_sites`.`id` = `subquery`.`elementsSitesId`
INNER JOIN `craft_campaign_campaigns` `campaign_campaigns` ON `campaign_campaigns`.`id` = `subquery`.`elementsId`
INNER JOIN `craft_content` `content` ON `content`.`id` = `subquery`.`contentId`

Error Info: Array
(
    [0] => 23000
    [1] => 1052
    [2] => Column 'dateCreated' in where clause is ambiguous
)

↵
Caused by: PDOException
SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'dateCreated' in where clause is ambiguous

in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Command.php at line 1302

    1. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Schema.php at line 676
    667668669670671672673674675676677678679680681682683684685

     
            $exceptionClass = '\yii\db\Exception';
            foreach ($this->exceptionMap as $error => $class) {
                if (strpos($e->getMessage(), $error) !== false) {
                    $exceptionClass = $class;
                }
            }
            $message = $e->getMessage() . "\nThe SQL being executed was: $rawSql";
            $errorInfo = $e instanceof \PDOException ? $e->errorInfo : null;
            return new $exceptionClass($message, $errorInfo, $e->getCode(), $e);
        }
     
        /**
         * Returns a value indicating whether a SQL statement is for read purpose.
         * @param string $sql the SQL statement
         * @return bool whether a SQL statement is for read purpose.
         */
        public function isReadQuery($sql)
        {

    2. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Command.php at line 1307 – [yii\db\Schema::convertException](https://www.yiiframework.com/doc-2.0/yii-db-schema.html#convertException()-detail)(PDOException, 'SELECT COUNT(*) FROM (SELECT `ca...')
    3. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Command.php at line 1168 – [yii\db\Command::internalExecute](https://www.yiiframework.com/doc-2.0/yii-db-command.html#internalExecute()-detail)('SELECT COUNT(*) FROM (SELECT `ca...')
    4. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Command.php at line 436 – [yii\db\Command::queryInternal](https://www.yiiframework.com/doc-2.0/yii-db-command.html#queryInternal()-detail)('fetchColumn', 0)
    5. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Query.php at line 497 – [yii\db\Command::queryScalar](https://www.yiiframework.com/doc-2.0/yii-db-command.html#queryScalar()-detail)()
    6. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/db/Query.php at line 367 – [yii\db\Query::queryScalar](https://www.yiiframework.com/doc-2.0/yii-db-query.html#queryScalar()-detail)('COUNT(*)', null)
    361362363364365366367368369370371372

        /**
         * @inheritdoc
         */
        protected function queryScalar($selectExpression, $db): bool|string|null
        {
            try {
                return parent::queryScalar($selectExpression, $db);
            } catch (QueryAbortedException) {
                return false;
            }
        }
    }

    7. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/db/Query.php at line 368 – craft\db\Query::queryScalar('COUNT(*)', null)
    8. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/elements/db/ElementQuery.php at line 1549 – [yii\db\Query::count](https://www.yiiframework.com/doc-2.0/yii-db-query.html#count()-detail)('*', null)
    1543154415451546154715481549155015511552155315541555

        {
            // Cached?
            if (($cachedResult = $this->getCachedResult()) !== null) {
                return count($cachedResult);
            }
     
            return parent::count($q, $db) ?: 0;
        }
     
        /**
         * @inheritdoc
         * @return ElementInterface[]|array
         */

    9. in /home/forge/imagescience.com.au/releases/dev/vendor/putyourlightson/craft-campaign/src/widgets/CampaignStatsWidget.php at line 126 – craft\elements\db\ElementQuery::count()
    120121122123124125126127128129130131132

                $contactCampaignQuery->andWhere(['and',
                    ['>=', 'sent', $startDate],
                    ['<', 'sent', $endDate],
                ]);
            }
     
            $campaigns = $campaignQuery->count();
            $sendouts = $sendoutQuery->count();
            $recipients = (clone $contactCampaignQuery)->count();
            $opened = (clone $contactCampaignQuery)->andWhere(['not', ['opened' => null]])->count();
            $clicked = (clone $contactCampaignQuery)->andWhere(['not', ['clicked' => null]])->count();
     
            return Craft::$app->getView()->renderTemplate('campaign/_widgets/campaign-stats/widget', [

    10. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/controllers/DashboardController.php at line 466 – putyourlightson\campaign\widgets\CampaignStatsWidget::getBodyHtml()
    460461462463464465466467468469470471472

         */
        private function _getWidgetInfo(WidgetInterface $widget): array|false
        {
            $view = $this->getView();
     
            // Get the body HTML
            $widgetBodyHtml = $widget->getBodyHtml();
     
            if ($widgetBodyHtml === null) {
                return false;
            }
     
            // Get the settings HTML + JS

    11. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/controllers/DashboardController.php at line 104 – craft\controllers\DashboardController::_getWidgetInfo(putyourlightson\campaign\widgets\CampaignStatsWidget)
    9899100101102103104105106107108109110

            $variables['widgets'] = [];
            $widgets = $dashboardService->getAllWidgets();
            $allWidgetJs = '';
     
            foreach ($widgets as $widget) {
                $view->startJsBuffer();
                $info = $this->_getWidgetInfo($widget);
                $widgetJs = $view->clearJsBuffer(false);
     
                if ($info === false) {
                    continue;
                }
     

    12. craft\controllers\DashboardController::actionIndex()
    13. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/base/InlineAction.php at line 57 – call_user_func_array([craft\controllers\DashboardController, 'actionIndex'], [])
    14. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/base/Controller.php at line 178 – [yii\base\InlineAction::runWithParams](https://www.yiiframework.com/doc-2.0/yii-base-inlineaction.html#runWithParams()-detail)([])
    15. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/base/Module.php at line 552 – [yii\base\Controller::runAction](https://www.yiiframework.com/doc-2.0/yii-base-controller.html#runAction()-detail)('index', [])
    16. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/web/Application.php at line 304 – [yii\base\Module::runAction](https://www.yiiframework.com/doc-2.0/yii-base-module.html#runAction()-detail)('dashboard/index', [])
    298299300301302303304305306307308309310

         * @param string $route
         * @param array $params
         * @return Response|null The result of the action, normalized into a Response object
         */
        public function runAction($route, $params = []): ?BaseResponse
        {
            $result = parent::runAction($route, $params);
     
            if ($result === null || $result instanceof Response) {
                return $result;
            }
     
            $response = $this->getResponse();

    17. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/web/Application.php at line 103 – craft\web\Application::runAction('dashboard/index', [])
    18. in /home/forge/imagescience.com.au/releases/dev/vendor/craftcms/cms/src/web/Application.php at line 289 – [yii\web\Application::handleRequest](https://www.yiiframework.com/doc-2.0/yii-web-application.html#handleRequest()-detail)(craft\web\Request)
    283284285286287288289290291292293294295

            if (($response = $this->_processActionRequest($request)) !== null) {
                return $response;
            }
     
            // If we’re still here, finally let Yii do its thing.
            try {
                return parent::handleRequest($request);
            } catch (Throwable $e) {
                $this->_unregisterDebugModule();
                throw $e;
            }
        }
     

    19. in /home/forge/imagescience.com.au/releases/dev/vendor/yiisoft/yii2/base/Application.php at line 384 – craft\web\Application::handleRequest(craft\web\Request)
    20. in /home/forge/imagescience.com.au/releases/dev/web/index.php at line 12 – [yii\base\Application::run](https://www.yiiframework.com/doc-2.0/yii-base-application.html#run()-detail)()
    6789101112

    // Load shared bootstrap
    require dirname(__DIR__) . '/bootstrap.php';
     
    // Load and run Craft
    /** @var craft\web\Application $app */
    $app = require CRAFT_VENDOR_PATH . '/craftcms/cms/bootstrap/web.php';
    $app->run();
    ```

**Comment by bossanova808** [comment]
(I deleted the rows from the `craft_widgets` table and can load the dashboard again...but obviously it shouldn't get into a state where the first page of the control panel won't load at all...)

This is latest Craft 4 & Campaign

**Comment by bencroker** [comment]
Thanks for reporting that! Fixed in https://github.com/putyourlightson/craft-campaign/commit/3271dd3c8b7c54f14ce157d9ce5f2650f12b08b9 and released in version 2.8.2.
